title: DiffBind Chrissy Version
Author: CBREEZY
date: 12/13/2022
output:html_document
objective: PART 1 - load in .bed and .bam files from each CUT&Tag samples and get the consensus peaks for each replicate, then wanted output is RPKM files

```{r load library}
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("DiffBind")
#install.packages("amap", type = "source")
#library(amap)
#install.packages("mixsqp", type = "binary")
#library(mixsqp)
# install.packages("devtools")
# install.packages("ellipsis")
#library("ellipsis")
#library(devtools)
# install.packages("mixsqp", )
# install_github("stephenslab/mixsqp",build_vignettes = TRUE)
# install.packages("DiffBind", type = "binary")
#if (!require("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")

#BiocManager::install("profileplyr")
#install.packages("profileplyr")
library(profileplyr)
library(tidyverse)
library(DiffBind)
library(gplots)
```


```{r Call the data/create DBA object, H3K4Me3 Analysis}
setwd("/Users/christinenguyen/Desktop/diffbind")

#Create Sample Sheet with BAM and BED files, how to write the samplesheet: https://rdrr.io/bioc/DiffBind/man/dba.html
#if run into "incomplete final line found by readTableHeader, go directly to samplesheet, open via textEdit, click last line and press enter, https://www.statology.org/r-incomplete-final-line-found-by-readtableheader/
H3K4Me3 <-dba(sampleSheet = "diffbind.csv", minOverlap = 2) 
H3K4Me3
#4 Samples, 23753 sites in matrix, after using mm10 .bed file as the Peak
#4 Samples, 3572 sites (that overlap in at least 2 samples) in matrix (371116 total UNIQUE PEAKS)
# #This shows how many peaks are in each peakset, as well as (in the first line) the total number
# of unique peaks after merging overlapping ones (371116), and the dimensions of the default
# binding matrix of 4 samples by the 3572 sites that overlap in at least two of the samples.

dba.peakset(H3K4Me3, bRetrieve = TRUE)

#Occcupancy
olap.rate <-dba.overlap(H3K4Me3, mode = DBA_OLAP_RATE)
olap.rate
plot(olap.rate)
#23753 23753 23753 23753

#how well do replicates agree on their peak calls?
names(H3K4Me3$masks)
dba.overlap(H3K4Me3, H3K4Me3$masks$`CXCR6+`, mode = DBA_OLAP_RATE)
#23753 23753

dba.overlap(H3K4Me3, H3K4Me3$masks$`CXCR5+`, mode = DBA_OLAP_RATE)
#23753 out of 23753 peaks identified in 2 cxcr5 replicates

H3K4Me3_consensus <- dba.peakset(H3K4Me3, consensus = DBA_CONDITION, minOverlap = 0.33) #0.33 and 0.66 is no difference, 0.33 actually combines the samples vs also including the replicates in the analysis
H3K4Me3_consensus

H3K4Me3_consensus

H3K4Me3_consensus <-dba(H3K4Me3_consensus, mask = H3K4Me3_consensus$masks$Consensus, minOverlap = 1)
H3K4Me3_consensus

consensus_peaks <- dba.peakset(H3K4Me3_consensus, bRetrieve = TRUE)
consensus_peaks

H3K4Me3.OL <- dba.overlap(H3K4Me3_consensus, H3K4Me3_consensus$masks$Consensus)
H3K4Me3.OL

#Th1 only
H3K4Me3.OL$onlyA

#Tfh only
H3K4Me3.OL$onlyB

#all H3K4Me3
H3K4Me3.OL$inAll

```


```{r Create .bed files of consensus peaks, h3k4me3}

#Th1_H3K4Me3_unique
Th1_H3K4Me3_unique <- H3K4Me3.OL$onlyA
write.table(Th1_H3K4Me3_unique, file = "Th1_H3K4Me3_unique.bed",
            sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)

#Tfh_H3K4Me3_unique
Tfh_H3K4Me3_unique <- H3K4Me3.OL$onlyB
write.table(Tfh_H3K4Me3_unique, file = "Tfh_H3K4Me3_unique.bed",
            sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)

#Shared_H3K4Me3
Shared_H3K4Me3 <- H3K4Me3.OL$inAll
write.table(Shared_H3K4Me3, file = "Shared_H3K4Me3.bed",
            sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)

```

```{r this is where u can run dba.COUNT, then get RPKM reads for all genes, H3K4Me3}
#run this for affinity binding analysis only, if u run this after the dba.peakset, you cant go back
H3K4Me3_count <- dba.count(H3K4Me3, peaks = consensus_peaks)
H3K4Me3_count

H3K4Me3_count <- dba.count(H3K4Me3, bUseSummarizeOverlaps = FALSE)
H3K4Me3_count

H3K4Me3_count <- dba.count(H3K4Me3, peaks = "./mm10.tss.bed")
H3K4Me3_count

H3K4Me3_RPKM_all <- tibble(Chr = H3K4Me3_count$peaks[[1]]$Chr, 
                           Start = H3K4Me3_count$peaks[[1]]$Start, 
                           End = H3K4Me3_count$peaks[[1]]$End, 
                           H3K4Me3_RPKM_CXCR6_1 = H3K4Me3_count$peaks[[1]]$RPKM, 
                           H3K4Me3_RPKM_CXCR6_2 = H3K4Me3_count$peaks[[2]]$RPKM, 
                           H3K4Me3_RPKM_CXCR5_1 = H3K4Me3_count$peaks[[3]]$RPKM, 
                           H3K4Me3_RPKM_CXCR5_2 = H3K4Me3_count$peaks[[4]]$RPKM)

write_delim(H3K4Me3_RPKM_all, "H3K4Me3_RPKM_all.txt", delim = "\t")
```

```{r Call the data/create DBA object, H3K27Ac Analysis}
setwd("/Users/christinenguyen/Desktop/diffbind")
#H3K4Me3 is H3K4Me3 LMAO
#Create Sample Sheet with BAM and BED files, how to write the samplesheet: https://rdrr.io/bioc/DiffBind/man/dba.html
#if run into "incomplete final line found by readTableHeader, go directly to samplesheet, open via textEdit, click last line and press enter, https://www.statology.org/r-incomplete-final-line-found-by-readtableheader/
H3K27Ac <-dba(sampleSheet = "diffbind_Samplesheet_H3K27Ac2.csv", minOverlap = 1) 
H3K27Ac
#4 Samples, 1869 sites (that overlap in at least 2 samples) in matrix (1291666 total UNIQUE PEAKS)
# #This shows how many peaks are in each peakset, as well as (in the first line) the total number
# of unique peaks after merging overlapping ones (1291666), and the dimensions of the default
# binding matrix of 4 samples by the 1869 sites that overlap in at least two of the samples.

#Occcupancy
olap.rate <-dba.overlap(H3K27Ac, mode = DBA_OLAP_RATE)
olap.rate
plot(olap.rate)
#1728 446

#how well do replicates agree on their peak calls?
names(H3K27Ac$masks)
dba.overlap(H3K27Ac, H3K27Ac$masks$`CXCR6`, mode = DBA_OLAP_RATE)
#504 out of 1686 identified in 2 cxcr6 replicates
#1201, 1 sample

dba.overlap(H3K27Ac, H3K27Ac$masks$`CXCR5`, mode = DBA_OLAP_RATE)
#1085 out of 1292675 peaks identified in 2 cxcr5 replicates
#1144, 1 sample

dba.plotVenn(H3K27Ac, H3K27Ac$masks$'CXCR6+')

H3K27Ac_consensus <- dba.peakset(H3K27Ac, consensus = DBA_CONDITION, minOverlap = 0.33) #0.33 and 0.66 is no difference, 0.33 actually combines the samples vs also including the replicates in the analysis
H3K27Ac_consensus <- dba.peakset(H3K27Ac, consensus = DBA_CONDITION)
H3K27Ac_consensus

H3K27Ac_consensus <-dba(H3K27Ac_consensus, mask = H3K27Ac_consensus$masks$Consensus, minOverlap = 1)
H3K27Ac_consensus

consensus_peaks2 <- dba.peakset(H3K27Ac_consensus, bRetrieve = TRUE)
consensus_peaks2

H3K27Ac.OL <- dba.overlap(H3K27Ac, H3K27Ac_consensus$masks$Consensus)
H3K27Ac.OL

#Th1 only
H3K27Ac.OL$onlyA

#Tfh only
H3K27Ac.OL$onlyB

#all H3K4Me3
H3K27Ac.OL$inAll

```

```{r Create .bed files of consensus peaks, h3k27ac}

#Th1_H3K27Ac_unique
Th1_H3K27Ac_unique <- H3K27Ac.OL$onlyA
write.table(Th1_H3K27Ac_unique, file = "Th1_H3K27Ac_unique.bed",
            sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)

#Tfh_H3K27Ac_unique
Tfh_H3K27Ac_unique <- H3K27Ac.OL$onlyB
write.table(Tfh_H3K27Ac_unique, file = "Tfh_H3K27Ac_unique.bed",
            sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)

#Shared_H3K27Ac
Shared_H3K27Ac <- H3K27Ac.OL$inAll
write.table(Shared_H3K27Ac, file = "Shared_H3K27Ac.bed",
            sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)

```

```{r this is where u can run dba.COUNT, then get RPKM reads for all genes, H3K27Ac}
#run this for affinity binding analysis only, if u run this after the dba.peakset, you cant go back
H3K27Ac_count <- dba.count(H3K27Ac, peaks = consensus_peaks2)
H3K27Ac_count

H3K27Ac_RPKM_all <- tibble(Chr = H3K27Ac_count$peaks[[1]]$Chr, 
                           Start = H3K27Ac_count$peaks[[1]]$Start, 
                           End = H3K27Ac_count$peaks[[1]]$End, 
                           H3K27Ac_RPKM_CXCR6_1 = H3K27Ac_count$peaks[[1]]$RPKM, 
                           H3K27Ac_RPKM_CXCR6_2 = H3K27Ac_count$peaks[[2]]$RPKM, 
                           H3K27Ac_RPKM_CXCR5_1 = H3K27Ac_count$peaks[[3]]$RPKM, 
                           H3K27Ac_RPKM_CXCR5_2 = H3K27Ac_count$peaks[[4]]$RPKM)

write_delim(H3K27Ac_RPKM_all, "H3K27Ac_RPKM_all.txt", delim = "\t")

```

```{r Call the data/create DBA object, H3K27Me3 Analysis}
setwd("/Users/christinenguyen/Desktop/diffbind")
#H3K4Me3 is H3K4Me3 LMAO
#Create Sample Sheet with BAM and BED files, how to write the samplesheet: https://rdrr.io/bioc/DiffBind/man/dba.html
#if run into "incomplete final line found by readTableHeader, go directly to samplesheet, open via textEdit, click last line and press enter, https://www.statology.org/r-incomplete-final-line-found-by-readtableheader/
H3K27Me3 <-dba(sampleSheet = "diffbind_Samplesheet_H3K27Me3.csv", minOverlap = 2) 
H3K27Me3
#4 Samples, 5398 sites (that overlap in at least 2 samples) in matrix (2021201 total UNIQUE PEAKS)
# #This shows how many peaks are in each peakset, as well as (in the first line) the total number
# of unique peaks after merging overlapping ones (2021201), and the dimensions of the default
# binding matrix of 4 samples by the 5398 sites that overlap in at least two of the samples.

#Occcupancy
olap.rate <-dba.overlap(H3K27Me3, mode = DBA_OLAP_RATE)
olap.rate
plot(olap.rate)
#2021201    5398    3338    1863

#how well do replicates agree on their peak calls?
names(H3K27Me3$masks)
dba.overlap(H3K27Me3, H3K27Me3$masks$`CXCR6`, mode = DBA_OLAP_RATE)
#3416 out of 3416 identified in 1 cxcr6 replicates

dba.overlap(H3K27Me3, H3K27Me3$masks$`CXCR5`, mode = DBA_OLAP_RATE)
#4017 out of 2031712 peaks identified in 2 cxcr5 replicates

H3K27Me3_consensus <- dba.peakset(H3K27Me3, consensus = DBA_CONDITION, minOverlap = 0.33) #0.33 and 0.66 is no difference, 0.33 actually combines the samples vs also including the replicates in the analysis
H3K27Me3_consensus

#recreate the object with the consensus peaks
H3K27Me3_consensus <-dba(H3K27Me3_consensus, mask = H3K27Me3_consensus$masks$Consensus, minOverlap = 1)
H3K27Me3_consensus

consensus_peaks3 <- dba.peakset(H3K27Me3_consensus, bRetrieve = TRUE)
consensus_peaks3

H3K27Me3.OL <- dba.overlap(H3K27Me3_consensus, H3K27Me3_consensus$masks$Consensus)
H3K27Me3.OL

#Th1 only
H3K27Me3.OL$onlyA

#Tfh only
H3K27Me3.OL$onlyB

#all H3K4Me3
H3K27Me3.OL$inAll

```

```{r Create .bed files of consensus peaks, h3k27me3}

#Th1_H3K27Me3_unique
Th1_H3K27Me3_unique <- H3K27Me3.OL$onlyA
write.table(Th1_H3K27Me3_unique, file = "Th1_H3K27Me3_unique.bed",
            sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)

#Tfh_H3K27Me3_unique
Tfh_H3K27Me3_unique <- H3K27Me3.OL$onlyB
write.table(Tfh_H3K27Me3_unique, file = "Tfh_H3K27Me3_unique.bed",
            sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)

#Shared_H3K27Me3
Shared_H3K27Me3 <- H3K27Me3.OL$inAll
write.table(Shared_H3K27Me3, file = "Shared_H3K27Me3.bed",
            sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)

```

```{r this is where u can run dba.COUNT, then get RPKM reads for all genes, H3K27Me3}
#run this for affinity binding analysis only, if u run this after the dba.peakset, you cant go bMe3k
H3K27Me3_count <- dba.count(H3K27Me3)
H3K27Me3_count

H3K27Me3_RPKM_all <- tibble(Chr = H3K27Me3_count$peaks[[1]]$Chr, 
                           Start = H3K27Me3_count$peaks[[1]]$Start, 
                           End = H3K27Me3_count$peaks[[1]]$End, 
                           H3K27Me3_RPKM_CXCR6_1 = H3K27Me3_count$peaks[[1]]$RPKM, 
                           H3K27Me3_RPKM_CXCR5_1 = H3K27Me3_count$peaks[[2]]$RPKM, 
                           H3K27Me3_RPKM_CXCR5_2 = H3K27Me3_count$peaks[[3]]$RPKM)

write_delim(H3K27Me3_RPKM_all, "H3K27Me3_RPKM_all.txt", delim = "\t")

```
